source/man/kirbydownload.1
'\" t
.\"     Title: Kirbydownload
.\"    Author: Uwe Gehring <adspectus@fastmail.com>
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 06/02/2020
.\"    Manual: kirbydownload User Manual
.\"    Source: kirbydownload
.\"  Language: English
.\"
.TH "KIRBYDOWNLOAD" "1" "06/02/2020" "kirbydownload" "kirbydownload User Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
kirbydownload \- Script to select and download a Kirby kit
.SH "SYNOPSIS"
.HP \w'\fBkirbydownload\fR\ 'u
\fBkirbydownload\fR [\fB\-d\fR] [\fB\-l\fR] [\fB\-f\fR] [\fB\-k\fR\ {\fBstarterkit\fR\ |\ \fBplainkit\fR}] [\fB\-v\fR\ {\fBcurrent\fR\ |\ \fIversionnumber\fR}] [\fB\-t\fR\ \fItargetdir\fR]
.HP \w'\fBkirbydownload\fR\ 'u
\fBkirbydownload\fR [\fB\-h\fR]
.SH "DESCRIPTION"
.PP
This manual page documents briefly the
\fBkirbydownload\fR
command, which is part of the
\fBkirbytools\fR(1)
package\&.
.PP
\fBkirbydownload\fR
is a script that will download a Kirby package from Github to your local machine\&. You can choose between the starterkit and the plainkit package and you can select the version which should be downloaded\&. You can also specify the directory where the package will be downloaded to\&.
.PP
Optionally, you can force the download even if a package of the same kit and version is already on your local machine, and you can ask to extract and install only the
kirby
program directory to a specific directory used for symbolic links of the same directory in a given virtual host installations directory\&.
.PP
Default values for these options will be set by the
\fBkirbyconfigure\fR(1)
script, which will be called upon first invocation of any of the
kirbytools
scripts\&. The values will be saved in the file
$HOME/\&.kirbyrc\&.
.SH "OPTIONS"
.PP
The script follows the usual GNU command line syntax, options always starting with a dash (`\-\*(Aq)\&.
.PP
\fB\-h\fR
.RS 4
Show summary of options\&.
.RE
.PP
\fB\-d\fR
.RS 4
Show debugging information\&.
.RE
.PP
\fB\-l\fR
.RS 4
Extract the
kirby
program directory of the downloaded kit to a folder with kirby program directories\&. The location defaults to the value of
\fBKIRBYLIBDIR\fR
(appended with the versionnumber), which has been configured by the
\fBkirbyconfigure\fR(1)
script and which is saved in the file
$HOME/\&.kirbyrc\&.
.sp
This makes sense if you would like to create multiple virtual hosts which are based on the same Kirby version\&. The
kirby
program directory will then only be linked into the virtual host folder\&. See the
\m[blue]\fBKirby cookbook recipe about multisite installations\fR\m[]\&\s-2\u[1]\d\s+2
for further details\&.
.RE
.PP
\fB\-f\fR
.RS 4
Force download even if the package is already downloaded\&. Standard behavior is to skip downloading if the package is already in
\fBKIRBYDOWNLOADDIR\fR\&.
.RE
.PP
\fB\-k \fR\fB\fBstarterkit\fR\fR\fB|\fR\fB\fBplainkit\fR\fR
.RS 4
Specify which Kirby
\fIkit\fR
to download\&. Defaults to the value of
\fBKIRBYKIT\fR, which has been configured by the
\fBkirbyconfigure\fR(1)
script and which is saved in the file
$HOME/\&.kirbyrc\&.
.RE
.PP
\fB\-v \fR\fB\fBcurrent\fR\fR\fB|\fR\fB\fIversionnumber\fR\fR
.RS 4
Specify which Kirby
\fIversion\fR
to download\&. Defaults to the value of
\fBKIRBYVERSION\fR, which has been configured by the
\fBkirbyconfigure\fR(1)
script and which is saved in the file
$HOME/\&.kirbyrc\&.
.sp
If
\fBcurrent\fR
is selected, always the most recent version is downloaded\&. Otherwise a valid versionnumber has to be specified\&.
.RE
.PP
\fB\-t \fR\fB\fItargetdir\fR\fR
.RS 4
Specify the directory where downloads will be saved\&. Defaults to the value of
\fBKIRBYDOWNLOADDIR\fR, which has been configured by the
\fBkirbyconfigure\fR(1)
script and which is saved in the file
$HOME/\&.kirbyrc\&.
.sp
The downloaded package is named
\fIpackage\fR\-\fIversion\fR\&.tar\&.gz\&. Hence, the full path would be
/usr/local/src/kirby/\fIpackage\fR\-\fIversion\fR\&.tar\&.gz
if the variable
\fBKIRBYDOWNLOADDIR\fR
points to
/usr/local/src/$KIRBYSUFFIX, and the variable
\fBKIRBYSUFFIX\fR
is defined as
\(lqkirby\(rq\&.
.RE
.SH "FILES"
.PP
/etc/kirbytools/kirbyrc
.RS 4
The system\-wide configuration file to control the behaviour of
kirbydownload\&.
.RE
.PP
${HOME}/\&.kirbyrc
.RS 4
The per\-user configuration file to control the behaviour of
kirbydownload\&. This file will be created by the
\fBkirbyconfigure\fR
script (see
\fBkirbyconfigure\fR(1)) if it does not exist\&.
.RE
.SH "BUGS"
.PP
Currently, only Kirby 3 starterkits and plainkits are supported\&.
.PP
The programs are currently optimized to work with the
apache2
httpd\&.
.PP
Some functionality requires root (or sudo as root) access\&.
.SH "SEE ALSO"
.PP
\fBkirbyconfigure\fR(1),
\fBkirbydeinstall\fR(1),
\fBkirbyinstall\fR(1),
\fBkirbysetup\fR(1),
\fBkirbytools\fR(1)
.SH "AUTHOR"
.PP
\fBUwe Gehring\fR <\&adspectus@fastmail\&.com\&>
.RS 4
.RE
.SH "COPYRIGHT"
.br
Copyright \(co 2020 Uwe Gehring
.br
.PP
This manual page was written for the Debian system (and may be used by others)\&.
.PP
Permission is granted to copy, distribute and/or modify this document under the terms of the GNU General Public License, Version 2 or (at your option) any later version published by the Free Software Foundation\&.
.PP
On Debian systems, the complete text of the GNU General Public License can be found in
/usr/share/common\-licenses/GPL\&.
.sp
.SH "NOTES"
.IP " 1." 4
Kirby cookbook recipe about multisite installations
.RS 4
\%https://getkirby.com/docs/cookbook/setup/multisite-variant
.RE


kirbytools_1.0/source/kirbyDownload.sh
#!/bin/bash
#
# kirbyDownload by Uwe Gehring <uwe@imap.cc>

## Read default variables
if [ -f /etc/kirbytools/kirbyrc ];then
  . /etc/kirbytools/kirbyrc
else
  echo "File /etc/kirbytools/kirbyrc not found!" && exit 2
fi

## Read default functions
if [ -f /etc/kirbytools/kirbyfunctions ];then
  . /etc/kirbytools/kirbyfunctions
else
  echo "File /etc/kirbytools/kirbyfunctions not found!" && exit 2
fi

## Initialize settings
initKirbySetup

## Usage
usage() {
  echo ""
  echo "  Usage:"
  echo ""
  echo "  $(basename $0) [-h] [-d] [-l] [-k <starterkit|plainkit>] [-v <current|versionnumber>] [-t <targetdir>]"
  echo ""
  echo "  -h: Show this help"
  echo "  -d: Show debug information"
  echo "  -l: Download and extract kirby dir to $KIRBYLIBDIR"
  echo "  -k: kit to be installed (default: ${txtbld}$KIRBYDEFAULTKIT${txtrst})"
  echo "  -v: version to be installed (default: ${txtbld}$KIRBYDEFAULTVERSION${txtrst})"
  echo "  -t: directory to be downloaded into (default: ${txtbld}$KIRBYDOWNLOADDIR${txtrst})"
  echo ""
  echo "  Current versionnumber for starterkit: ${txtbld}${_KIRBYTAGCURRENT[starterkit]}${txtrst}"
  echo "  Current versionnumber for plainkit: ${txtbld}${_KIRBYTAGCURRENT[plainkit]}${txtrst}"
  echo ""
  exit 1
}

## Get commandline options
while getopts ":dhlk:t:v:" opt;do
  case "${opt}" in
    d)  DEBUG=1;;
    h)  usage;;
    l)  LINKKIRBY=1;;
    k)  kit=$OPTARG;;
    t)  target=$OPTARG;;
    v)  version=$OPTARG;;
    :)  errMsg "-$OPTARG requires an argument" && usage;;
    *)  errMsg "Invalid option $OPTARG" && usage;;
  esac
done
shift $((OPTIND -1))

## Check if kit has valid value
kit=${kit:-$KIRBYDEFAULTKIT}
if [[ "$kit" != "starterkit" && "$kit" != "plainkit" ]];then
  errMsg "Invalid kit $kit" && usage
fi

## Check if version has valid value
version=${version:-$KIRBYDEFAULTVERSION}
if ! checkValidKirbyVersion $version $kit;then
  errMsg "Invalid version $version" && usage
fi
if [ "$version" == "current" ];then
  version=${_KIRBYTAGCURRENT[$kit]}
fi

## Check if target has valid value
target=${target:-$KIRBYDOWNLOADDIR}

## Print info in debug mode
[[ $DEBUG ]] && showVars

## Check first if kit is already downloaded. If not, get kit and save it
## in target dir and rename from version.tar.gz to kit-version.tar.gz
if [ ! -f $target/$kit-$version.tar.gz ];then
  [[ -z $KIRBYGITURL ]] && errMsg "KIRBYGITURL is not defined!" && exit 3
  sudo wget -q $KIRBYGITURL/$kit/archive/$version.tar.gz -P $target
  sudo mv $target/$version.tar.gz $target/$kit-$version.tar.gz
fi

## Unpack kirby directory to KIRBYLIBDIR
if [[ $LINKKIRBY && ! -d $KIRBYLIBDIR/$version ]];then
  tar -xzf $target/$kit-$version.tar.gz -C $KIRBYTEMPDIR
  mkdir -p $KIRBYLIBDIR/$version
  cp -au $KIRBYTEMPDIR/$kit-$version/kirby/. $KIRBYLIBDIR/$version
fi

## If all went right, print message and exit with code 0
echo -e "\n${txtgreen}Kirby $kit $version downloaded to $target/$kit-$version.tar.gz${txtrst}\n"

exit 0
## code: language=bash